modules/renter/worker.go

```diff
@@ -162,21 +178,107 @@ func (w *worker) upload(uw uploadWork) {
 		}
 	}
 	contract.Pieces = append(contract.Pieces, pieceData{
-		Chunk:      uw.chunkID.index,
-		Piece:      uw.pieceIndex,
+		Chunk:      uc.index,
+		Piece:      pieceIndex,
 		MerkleRoot: root,
 	})
-	uw.file.contracts[w.contractID] = contract
-	w.renter.saveFile(uw.file)
-	uw.file.mu.Unlock()
+	uc.renterFile.contracts[w.contractID] = contract
+	w.renter.saveFile(uc.renterFile)
+	uc.renterFile.mu.Unlock()
+
+	// Clear memory in the renter now that we're done.
+	w.renter.memoryAvailable += uint64(len(uc.physicalChunkData[pieceIndex]))
 	w.renter.mu.Unlock(id)
+	select {
+	case w.renter.newMemory <- struct{}{}:
+	default:
+	}
```

Comment:
I do think a `sync.Cond` is more appropriate here, but unfortunately there isn't a good way to select on both a `sync.Cond` and a shutdown channel. There's discussion of this problem [here](https://github.com/golang/go/issues/16620) but no clear solution forthcoming.

Reply:
Soon there will be two consumers. The uploader, and also the downloader. We can fix this though... with two channels... :>